external/kafka-0-10/src/main/scala/org/apache/spark/streaming/kafka010/CachedKafkaConsumer.scala

```diff
@@ -83,13 +81,50 @@ class CachedKafkaConsumer[K, V] private(
         s"Failed to get records for $groupId $topic $partition $offset after polling for $timeout")
       record = buffer.next()
       assert(record.offset == offset,
-        s"Got wrong record for $groupId $topic $partition even after seeking to offset $offset")
+        s"Got wrong record for $groupId $topic $partition even after seeking to offset $offset " +
+          s"got offset ${record.offset} instead. If this is a compacted topic, consider enabling " +
+          "spark.streaming.kafka.allowNonConsecutiveOffsets"
+      )
     }
 
     nextOffset = offset + 1
     record
   }
 
+  /**
+   * Start a batch on a compacted topic
+   */
+  def compactedStart(offset: Long, timeout: Long): Unit = {
+    logDebug(s"compacted start $groupId $topic $partition starting $offset")
+    // This seek may not be necessary, but it's hard to tell due to gaps in compacted topics
+    if (offset != nextOffset) {
+      logInfo(s"Initial fetch for compacted $groupId $topic $partition $offset")
+      seek(offset)
+      poll(timeout)
+    }
+  }
+
+  /**
+   * Get the next record in the batch from a compacted topic.
+   * Assumes compactedStart has been called first, and ignores gaps.
+   */
+  def compactedNext(timeout: Long): ConsumerRecord[K, V] = {
+    if (!buffer.hasNext()) { poll(timeout) }
+    assert(buffer.hasNext(),
```

Comment:
Should this really be an assert, like, should never happen regardless of input or external state? or just a condition that should generate an exception if false?

Reply:
That's a "shouldn't happen unless the topicpartition or broker is gone" kind of thing.  Semantically I could see that being more like require than assert, but don't have a strong opinion.

Reply:
`assert` turns into a JVM assert (I think?) and as such would be turned off if assertions are disabled, which is how it ought to run in production. If it's something that _could_ happen at all I think it should be `require`

Reply:
Agreed that require is better, will fix it in a sec

Pretty sure assert is just a function in Predef.scala that throws an AssertionError, it's not like a java assert statement that can be en / disabled with java -ea / -da.  Tested it out:

https://gist.github.com/koeninger/6155cd94a19d1a6373ba0b40039e97e3

Disabling scala asserts can be done at compile time with -Xdisable-assertions